### ISSUE TITLE
[Feature Request] Generalize @beartype validators to accept non-"bool" bool-like objects

### ISSUE THREAD

**Issue by braniii** [beartype/beartype#153]
Hi @leycec,
once again it's me. First, shame on me I forgot about #130. I put it back on my To-Do list Once-I-have-finished-my-first-thesis-draft :see_no_evil:, but sadly I can not promise when this will be.
Using my beartyped package MoSAIC some user got encountered with some non-userreadable error: `AssertionError: False not tri-state boolen`, moldyn/MoSAIC#9. I managed to reproduce this issue, even though I am not sure what is going wrong. So first a working example:
``` Python
import numpy as np
from beartype import beartype
from beartype.vale import Is
from beartype.typing import Annotated
from numpy.typing import NDArray

IsPositive = Is[lambda arr: np.all(arr > 0)]
IsSymmetric = Is[lambda arr: np.allclose(arr, arr.T)]

PositiveArray = Annotated[NDArray, IsPositive]
SymmetricArray = Annotated[NDArray, IsSymmetric]


@beartype
def is_symmetric(arr: SymmetricArray):
    print('arr is symmetric')


@beartype
def is_positive(arr: PositiveArray):
    print('arr is positive')


mat = np.array([[2, 1], [1, 2]])
is_symmetric(mat)
# arr is symmetric
is_positive(mat)
# arr is positive
```
Now making the matrix non symmetric throws the expected (and helpful) error
```Python
mat[0, 1] = 2
is_positive(mat)
# arr is positive
is_symmetric(mat)
# Traceback (most recent call last):
#   File "/home/user/Documents/beartype/issue153.py", line 29, in <module>
#     is_symmetric(mat)
#   File "<@beartype(__main__.is_symmetric) at 0x7f1b911da710>", line 28, in is_symmetric
#   File "/home/user/anaconda3/envs/beartype/lib/python3.10/site-packages/beartype/_decor/_error/errormain.py", line 301, in raise_pep_call_exception
#     raise exception_cls(  # type: ignore[misc]
# beartype.roar.BeartypeCallHintParamViolation: @beartyped is_symmetric() parameter arr="array([[2, 2],
#        [1, 2]])" violates type hint typing.Annotated[numpy.ndarray[typing.Any, numpy.dtype[+ScalarType]], Is[lambda arr: np.allclose(arr, arr.T)]], as "array([[2, 2],
#        [1, 2]])" violates validator Is[lambda arr: np.allclose(arr, arr.T)]:
#     False == Is[lambda arr: np.allclose(arr, arr.T)].
```
If we now use a non-positive matrix we will not be roared at by the bear, but get an AssertionError
```Python
mat[0, 1] = -2
mat[1, 0] = -2
is_symmetric(mat)
# arr is symmetric
is_positive(mat)
# Traceback (most recent call last):
#   File "/home/user/Documents/beartype/issue153.py", line 48, in <module>
#     is_positive(mat)
#   File "<@beartype(__main__.is_positive) at 0x7fad84de8a60>", line 28, in is_positive
#   File "/home/user/anaconda3/envs/beartype/lib/python3.10/site-packages/beartype/_decor/_error/errormain.py", line 292, in raise_pep_call_exception
#     ).get_cause_or_none()
#   File "/home/user/anaconda3/envs/beartype/lib/python3.10/site-packages/beartype/_decor/_error/_errorsleuth.py", line 344, in get_cause_or_none
#     return get_cause_or_none(self)
#   File "/home/user/anaconda3/envs/beartype/lib/python3.10/site-packages/beartype/_decor/_error/_pep/_errorpep593.py", line 87, in get_cause_or_none_annotated
#     hint_diagnosis = hint_validator.get_diagnosis(
#   File "/home/user/anaconda3/envs/beartype/lib/python3.10/site-packages/beartype/vale/_core/_valecore.py", line 425, in get_diagnosis
#     return format_diagnosis_line(
#   File "/home/user/anaconda3/envs/beartype/lib/python3.10/site-packages/beartype/vale/_util/_valeutiltext.py", line 73, in format_diagnosis_line
#     assert isinstance(is_obj_valid, (bool, type(None))), (
# AssertionError: False not tri-state boolean.
```
I have a creeping feeling that the fault is on my side, but I don't see it. It would be very nice if you could help me find it so that MoSAIC users don't get scared anymore.
Btw. I've learned from last time and executed the above code as a python script using the latest version of `beartype==10.4` and a not to old version of `numpy==1.21.6`.

Best
Daniel

**Comment by leycec** [comment]
**Yay!** It's my favourite molecular kinetics simulator lead dev. Back with another fearsome @beartype issue that roars explosively, @braniii in the house! :house: 

**Okay.** So, this is *sorta* a `numpy` issue and *sorta* a @beartype issue. On the one hand, our exception message here is clearly horrifying end users. That's bad and absolutely our fault. On the other hand, the `numpy.all()` function returns a `numpy`-specific `numpy.bool_` object rather than a standard `bool` object. That's kinda also bad and absolutely not our fault, but perhaps @beartype can and should transparently treat `numpy.bool_` objects as if they were just standard `bool` objects.

Until the next stable release of @beartype drops, MoSAIC (*...awesome stuff there, by the way – so excited for your upcoming thesis drop!*) can circumvent this on its end by explicitly coercing `numpy.bool_` objects into standard `bool` objects: e.g.,

```python
IsPositive = Is[lambda arr: bool(np.all(arr > 0))]
```

Thanks so much for exploiting @beartype for all it's worth. So much hugs. :hugs: :people_hugging:

**Comment by braniii** [comment]
@leycec It's so impressive, no matter what problem I have, you present such a simple fix :exploding_head: thx for that :)

**Comment by leycec** [comment]
Oh, you're most welcome. When you fly the friendly skies with @beartype, you just know you're in good *paws.* <sup>:smile: :laughing: :rofl:</sup>

Very well. I'll stop now. Luckily for everyone, the glorious puns end now! This has now been **resolved by commit 7e81f0f086146cc,** which substantially improves our exception handling to account for **bool-like validators** (i.e., functions subscripting `beartype.vale.Is[...]` that return objects that are *not* strictly booleans but *are* meaningfully convertible into booleans).

This was surprisingly non-trivial. Since you want to know more, I will now pontificate. For unknown reasons that I'd rather not think too hard about on a Friday night, *everything* in Python is convertible into booleans – just not necessarily meaningfully. Behold this grotesquery:

```python
>>> object OhGodsWhatsHappening(object): pass
>>> bool(OhGodsWhatsHappening())
True  # <-- WHAT THE HECK IS THIS, GUIDO. SRSLY, BRO. SRSLY.
```

So. We have established that the `bool()` constructor implicitly converts objects whose types do *not* define the `__bool__()` dunder method to ``True``. That's semantically insane and not at all how any other type coercions are handled. Imagine if the `int()` constructor implicitly converted objects whose types do *not* define the `__int__()` dunder method to **-42**, for example. This is like that.

But you're in good *paws* with @beartype. We hack around that nonsense and do The Right Thing™ the hard way. Why? Because we care. Okay. It's mostly because I'm chronically obsessed with type correctness. In any other world, that would be a personality disorder. Thankfully, I live in this world instead.

As always, thanks so much for your gratuitous interest in type-checking done the hard way. May your thesis be blessed with showering praise, grant funding, and the inevitable tenure tract we all know is coming your way. :man_teacher:
